### ISSUE TITLE
[flex_attention] Softcap perf questions

### ISSUE THREAD

**Issue by meshtag** [pytorch-labs/attention-gym#22]
I am using [this pytorch provided script](https://github.com/pytorch/pytorch/blob/main/benchmarks/transformer/score_mod.py) to benchmark flex attention with eager and got the attached results ([default_results.txt](https://github.com/user-attachments/files/16633693/default_results.txt)) on an A100. 

I modified the script to change the `score_mod` function to [`soft_cap` function ](https://pytorch.org/blog/flexattention/#soft-capping) like [this](https://gist.github.com/meshtag/936e170e85b5957124616e7471913679) and got the following results. Full result output is attached as well ([softcap_results.txt](https://github.com/user-attachments/files/16633694/softcap_results.txt)). Speedup is compared with eager (afaict).

```
=========================================================FWD Speedups========================================================


| Type    |   Speedup | score_mod   | mask_mod   | dtype          | shape(B,Hq,M,Hkv,N,D)       |
|---------|-----------|-------------|------------|----------------|-----------------------------|
| Average |     0.488 |             |            |                |                             |
| Max     |     0.537 | soft_cap    | None       | torch.bfloat16 | (2, 16, 512, 2, 512, 128)   |
| Min     |     0.385 | soft_cap    | None       | torch.bfloat16 | (2, 16, 1024, 16, 1024, 64) |
``` 

Is this the expected performance for softcap? Did I miss something? 

GPU used: A100
Packages and their versions 
```
torch                    2.5.0.dev20240812+cu118
pytorch-triton           3.0.0+dedb7bdf33
```

nvcc version: 
```
$ nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Wed_Sep_21_10:33:58_PDT_2022
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0
```
Please feel free to let me know if I need to provide more information. Thanks.

**Comment by joydddd** [comment]
You might want to use approx tanh instead of torch.tanh() for soft capping. check out this example: attn_gym/mods/softcapping.py

**Comment by Chillee** [comment]
Yes, definitely, you need to use approx tanh (like in this example: https://github.com/pytorch-labs/attention-gym/blob/5e0d1b8053a19b339ddf7c8015f2c4e02bf5b92c/attn_gym/mods/softcapping.py#L13). Normal `tanh` in CUDA is extremely expensive.

**Comment by SmerkyG** [comment]
Isn't the formula for softcapping incorrect? 

```py
    def tanh_softcap(score, b, h, q_idx, kv_idx):
        return score * tanh(score / soft_cap)
```

https://github.com/pytorch-labs/attention-gym/blob/5e0d1b8053a19b339ddf7c8015f2c4e02bf5b92c/attn_gym/mods/softcapping.py#L68

Seems like it should be 
```py
    def tanh_softcap(score, b, h, q_idx, kv_idx):
        return soft_cap * tanh(score / soft_cap)
```

**Comment by Chillee** [comment]
@SmerkyG Yeah, that's right haha. Do you want to submit a PR :)

**Comment by SmerkyG** [comment]
Sorry, I don't even use the repo or have it downloaded, was just looking quick and noticed!

**Comment by drisspg** [comment]
fixed
